metaserver/src/main/scala/scala/meta/languageserver/search/SymbolIndex.scala

```diff
@@ -220,47 +267,81 @@ class SymbolIndex(
+      ).flatMap { _.lift(symbol) }
+    }
+
+    /** If `case class A(a: Int)` and there is no companion object, resolve
+     * `A` in `A(1)` to the class definition.
+     */
+    val caseClassCompanionToType: PartialFunction[Symbol, Symbol] = {
       case Symbol.Global(owner, Signature.Term(name)) =>
-        // If `case class A(a: Int)` and there is no companion object, resolve
-        // `A` in `A(1)` to the class definition.
-        Symbol.Global(owner, Signature.Type(name)) :: Nil
-      case Symbol.Multi(ss) =>
-        // If `import a.B` where `case class B()`, then
-        // resolve to either symbol, whichever has a definition.
-        ss
-      case Symbol.Global(
-          companion @ Symbol.Global(owner, signature),
-          Signature.Method("apply" | "copy", _)
-          ) =>
-        // If `case class Foo(a: Int)`, then resolve
-        // `apply` in `Foo.apply(1)`, and
-        // `copy` in `Foo(1).copy(a = 2)`
-        // to the `Foo` class definition.
-        companion :: Symbol.Global(owner, Signature.Type(signature.name)) :: Nil
-      case Symbol.Global(owner, Signature.Method(name, _)) =>
-        Symbol.Global(owner, Signature.Term(name)) :: Nil
+        Symbol.Global(owner, Signature.Type(name))
+    }
+
+    /** If `case class Foo(a: Int)`, then resolve
+     * `a` in `Foo.apply(a = 1)`, and
+     * `a` in `Foo(1).copy(a = 2)`
+     * to the `Foo.a` primary constructor definition.
+     */
+    val caseClassApplyOrCopyParams: PartialFunction[Symbol, Symbol] = {
       case Symbol.Global(
           Symbol.Global(
             Symbol.Global(owner, signature),
             Signature.Method("copy" | "apply", _)
           ),
           param: Signature.TermParameter
           ) =>
-        // If `case class Foo(a: Int)`, then resolve
-        // `a` in `Foo.apply(a = 1)`, and
-        // `a` in `Foo(1).copy(a = 2)`
-        // to the `Foo.a` primary constructor definition.
         Symbol.Global(
           Symbol.Global(owner, Signature.Type(signature.name)),
           param
-        ) :: Nil
-      case _ =>
-        logger.info(s"Found no alternative for ${symbol.structure}")
-        Nil
+        )
+    }
+
+    /** If `case class Foo(a: Int)`, then resolve
+     * `apply` in `Foo.apply(1)`, and
+     * `copy` in `Foo(1).copy(a = 2)`
+     * to the `Foo` class definition.
+     */
+    val caseClassApplyOrCopy: PartialFunction[Symbol, Symbol] = {
+      case Symbol.Global(
+          Symbol.Global(owner, signature),
+          Signature.Method("apply" | "copy", _)
+          ) =>
+        Symbol.Global(owner, Signature.Type(signature.name))
+    }
+
+    // FIXME(alexey) I'm not sure what this transformation does
+    val methodToTerm: PartialFunction[Symbol, Symbol] = {
+      case Symbol.Global(owner, Signature.Method(name, _)) =>
+        Symbol.Global(owner, Signature.Term(name))
     }
+  }
+
+  /** A workaround for positions referring to jars */
+  def nonJarLocation(loc: l.Location): l.Location = {
```

Comment:
We can turn this into an extension method.

Reply:
Done in 36123b8bd71107152b187e2f607e2006afd7a6b9, although I don't like how it turned out. It uses `createFileInWorkspaceTarget` which needs `cwd`. IMO extension methods that needs too much context are not so nice. Anyway, we can move it around later.

Reply:
I agree an extension method should not require too much context, and should actually not do I/O either. It's fine for now however